### PR TITLE
Enable external resource monitoring

### PR DESCRIPTION
**Pull request by portertech**
Possible solution for https://github.com/sensu/sensu/issues/541

https://trello.com/c/eowfiJQ8/21-as-an-operator-i-would-like-to-monitor-external-resources-that-do-not-have-sensu-client-they-require-their-own-check-execution-h

### CODE REVIEW

**Comment by portertech** [comment]
Not sure about the use of `:keepalives => false`, may need a better attribute name to disable keepalive monitoring.

**Comment by portertech** [comment]
Ready for review/feedback :+1:

**Comment by portertech** [comment]
horsehead

File: lib/sensu/server/process.rb [hunk 1]
```diff
@@ -552,6 +586,7 @@ def determine_stale_clients
             @redis.get("client:#{client_name}") do |client_json|
               unless client_json.nil?
                 client = MultiJson.load(client_json)
+                next if client[:keepalives] == false
```
Comment (solarkennedy) [inline lib/sensu/server/process.rb]: This kinda hints that any JIT client created by an arbitrary check-source will not be auto-pruned (seems legit)
Works for me. This has the effect that every client ever will remain in the client list until manually deleted right? I'm ok with that if you are (not like it hurts anything)

**Comment by solarkennedy** [comment]
This looks good to me! I believe it will solve Yelp/sensu#1 !

I'm glad that you are using "source" as the name for the external client name key, as it makes it backwards compatible with Yelp's fork, but it also means that you... couldn't think of a better name either? :)

If you get this in mainline for a release, I'll volunteer to write docs?

File: lib/sensu/server/process.rb (hunk 1 above)
Comment (portertech) [inline lib/sensu/server/process.rb]: Correct, keepalive events will never be created for these clients, therefor, deregistration handlers will never be executed. Manually deleting a client will resolve current events and cleanup history etc \o/

**Comment by portertech** [comment]
@solarkennedy \o/ check "source" is the best I can do :P need to update the validation to use the client name regex tho, so there will be that constraint https://github.com/sensu/sensu-settings/blob/master/lib/sensu/settings/validators/check.rb#L72

**Comment by portertech** [comment]
@solarkennedy is Yelp using check sources with special characters and space? 

PR: https://github.com/sensu/sensu-settings/pull/8

**Comment by calebhailey** [comment]
This makes me happy. 

Thanks for your quick feedback @solarkennedy! :+1:

**Comment by mholtrlc** [comment]
I'm trying to understand how to make this work, I don't see much in terms of documentation of this feature.

**Comment by portertech** [comment]
@mholtrlc https://sensuapp.org/docs/latest/checks#definition-attributes check "source" which works with https://sensuapp.org/docs/latest/clients#jit-clients

**Comment by mholtrlc** [comment]
ok, I was on the right rack with the jit-clients but missed the source attribute.

this is my configuration and I don't see anything showing up in the dashboard

```
{
	"checks": {
		"source": {
			"source": "VC-SRV-SW-A",
			"command": "/bin/true",
			"subscribers": [
				"vc-srv-sw-a"
			],
			"interval": 1
		},
		"unpublished": {
			"command": "/bin/true",
			"publish": false,
			"subscribers": [
				"vc-srv-sw-a"
			]
  		},
	    "VC-SRV-SW-A_Ping": {
	      "type": "metric",
	      "command": "/etc/sensu/plugins/check-ping.rb -h 10.10.11.149 --report",
	      "subscribers": ["vc-srv-sw-a"],
	      "interval": 10
	      "source": "vc-srv-sw-a"
	    }

	}
}
```

**Comment by portertech** [comment]
@mholtrlc you are missing a comma after `"interval": 10`, invalid JSON, it will be logged and ignored.

**Comment by mholtrlc** [comment]
haha i've made that mistake more than once lately.  I fixed that but am still not seeing it in the dashboard but I see it in the logs.

````
{"timestamp":"2015-05-20T09:08:01.147803-0700","level":"info","message":"publishing check request","payload":{"name":"VC-SRV-SW-A_Ping","issued":1432138081,"command":"/etc/sensu/plugins/check-ping.rb -h 10.10.11.149 --report"},"subscribers":["vc-srv-sw-a"]}
{"timestamp":"2015-05-20T09:08:01.618088-0700","level":"info","message":"publishing check request","payload":{"name":"source","issued":1432138081,"command":"/bin/true"},"subscribers":["vc-srv-sw-a"]}
{"timestamp":"2015-05-20T09:08:02.644661-0700","level":"info","message":"publishing check request","payload":{"name":"source","issued":1432138082,"command":"/bin/true"},"subscribers":["vc-srv-sw-a"]}
````

**Comment by mholtrlc** [comment]
i'm thinking I don't have the correct configuration.  Here's what i've got:

client.json
````
{
  "client": {
    "name": "VC-SRV-SW-A",
    "address": "unknown",
    "subscriptions": [],
    "keepalives": false
  }
}
check.json
````
{
  "checks": {
    "VC-SRV-SW-A_Ping": {
      "source": "VC-SRV-SW-A",
      "command": "/etc/sensu/plugins/check-ping.rb -h 10.10.11.149 --report",
      "subscribers": [],
      "interval": 10
    }
  }
}

**Pull request merged**